### DISCUSSION TITLE
Does mold supports .gnu.hash ?

### DISCUSSION THREAD

**Discussion by LifeIsStrange** [rui314/mold#307] [Q&A] [answered]
.gnu.hash
I guess that the linker should set this by default --hash-style SHT_GNU_HASH
cf: https://flapenguin.me/elf-dt-gnu-hash

**Comment by rui314** [comment, marked as answer]
Yes it does. It's enabled by default.

**Reply by LifeIsStrange** [reply]
@rui314 
Thx, BTW your avatar is really cool :)
I suspect ELF to have some suboptimal implementations and for example GNU hash bring a major performance improvements but it is extremely likely that the gnu hashmap itself is no longer competitive.
Swisstable (and hashbrown) is the fastest hashmap on earth and using it would lead to significantly faster linking time and to much faster symbol lookup performance which is a cornerstone for the possibilities of debugging (e.g reverse/time travel debugging, omniscient debugging) and its performance on large scale projects.
As such i truly believe ELF (controlled by the binutils?) should integrate or link to Swisstable when available (or when embedded)
https://github.com/taviso/swisstable
If you find the project worth pursuing, you might be the only dev on earth that would simultaneously see the issue and have the required low level skill/erudition in order to extends ELF support for this much faster hashmap.

**Reply by rui314** [reply]
I don't think the performance of the .gnu.hash matters much, because .gnu.hash does not contain only a on-disk hash table but also a bloom filter along with it. The bloom filter is used to quickly filter out symbols that have zero chance to be provided by the shared object file. So the likelihood to hit a hash table and get a negative (not found) result is very low. I think it's like 5%, and 95% symbol lookups are filtered out by the bloom filter.
